lib/plugins/place_entity.js

```diff
@@ -0,0 +1,175 @@
+const assert = require('assert')
+const { callbackify } = require('../promise_utils')
+
+module.exports = inject
+
+function inject (bot, { version }) {
+  const Item = require('prismarine-item')(version)
+
+  async function placeEntityWithOptions (referenceBlock, faceVector, options) {
+    if (!bot.heldItem) throw new Error('must be holding an item to place an entity')
+
+    const type = bot.heldItem.name // used for assert
+      .replace(/.+_boat/, 'boat')
+      .replace(/.+_spawn_egg/, 'spawn_egg')
+    assert(['end_crystal', 'boat', 'spawn_egg', 'armor_stand'].includes(type), 'Unimplemented')
+
+    let name = bot.heldItem.name // used for finding entity after spawn
+      .replace(/.+_boat/, 'boat')
+
+    if (name.endsWith('spawn_egg')) {
+      name = bot.heldItem.spawnEggMobName
+    }
+    // Look at the center of the face
+    let dx = 0.5 + faceVector.x * 0.5
```

Comment:
is there any code from all this you could put in common with place block ? looks like yes

Reply:
most of the place logic is the same, yeah

Reply:
so try to do a function with the common stuff and call it here and in place block
will decrease the maintenance cost

Reply:
Done